<a href="https://colab.research.google.com/github/itsGab/previsao_uti_em_covid/blob/main/notebooks/tratamento/papc_tratamento.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<h1 align='center'>Predição de Admissão por COVID-19</h1>
<h2 align='center'>Tratamento e Limpeza dos Dados</h2>

# Abrindo os dados e leituras iniciais

In [3]:
data = "https://github.com/itsGab/previsao_uti_em_covid/blob/main/dados/Kaggle_Sirio_Libanes_ICU_Prediction.xlsx?raw=true"

In [4]:
import pandas as pd
df_inicial = pd.read_excel(data)
df_inicial.head()

PATIENT_VISIT_IDENTIFIER  AGE_ABOVE65  ...    WINDOW  ICU
0                         0            1  ...       0-2    0
1                         0            1  ...       2-4    0
2                         0            1  ...       4-6    0
3                         0            1  ...      6-12    0
4                         0            1  ...  ABOVE_12    1

[5 rows x 231 columns]



Separação de colunas por tipo do dados:

| nome da variável | descrição |
|---|---|
| cols_comorbidades | grupos de doenças |
| cols_demograficos | faixa etária e sexo |
| cols_sinais_vitais| sinais vitais |
| cols_exames_lab | exames laboratoriais | 
| cols_outros | identificador, janela e uti |

In [6]:
# baseado em um notebook feito pelo criador do desafio
# https://www.kaggle.com/felipeveiga/starter-covid-19-sirio-libanes-icu-admission

cols_comorbidades = [i for i in df_inicial.columns if "DISEASE" in i]
cols_comorbidades.extend(["HTN", "IMMUNOCOMPROMISED", "OTHER"])

cols_demografica = [i for i in df_inicial.columns if "AGE_" in i]
cols_demografica.append("GENDER")

cols_sinais_vitais = df_inicial.iloc[:,193:-2].columns.tolist()

cols_exames_lab = df_inicial.iloc[:,13:193].columns.tolist()

cols_outras = ["PATIENT_VISIT_IDENTIFIER", "WINDOW", "ICU"]

In [7]:
cols_demografica

['AGE_ABOVE65', 'AGE_PERCENTIL', 'GENDER']

In [ ]:
df_inicial.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1925 entries, 0 to 1924
Columns: 231 entries, PATIENT_VISIT_IDENTIFIER to ICU
dtypes: float64(225), int64(4), object(2)
memory usage: 3.4+ MB


In [ ]:
df_inicial.shape

(1925, 231)

In [ ]:
df_inicial['PATIENT_VISIT_IDENTIFIER'].nunique()

385

In [ ]:
pd.set_option('display.max_columns', None)
df_inicial.describe()

PATIENT_VISIT_IDENTIFIER  AGE_ABOVE65       GENDER  DISEASE GROUPING 1  \
count               1925.000000  1925.000000  1925.000000         1920.000000   
mean                 192.000000     0.467532     0.368831            0.108333   
std                  111.168431     0.499074     0.482613            0.310882   
min                    0.000000     0.000000     0.000000            0.000000   
25%                   96.000000     0.000000     0.000000            0.000000   
50%                  192.000000     0.000000     0.000000            0.000000   
75%                  288.000000     1.000000     1.000000            0.000000   
max                  384.000000     1.000000     1.000000            1.000000   

       DISEASE GROUPING 2  DISEASE GROUPING 3  DISEASE GROUPING 4  \
count         1920.000000         1920.000000         1920.000000   
mean             0.028125            0.097917            0.019792   
std              0.165373            0.297279            0.139320   
min              0.000000            0.000000            0.000000   
25%              0.000000            0.000000            0.000000   
50%              0.000000            0.000000            0.000000   
75%              0.000000            0.000000            0.000000   
max              1.000000            1.000000            1.000000   

       DISEASE GROUPING 5  DISEASE GROUPING 6          HTN  IMMUNOCOMPROMISED  \
count         1920.000000         1920.000000  1920.000000        1920.000000   
mean             0.128125            0.046875     0.213021           0.158333   
std              0.334316            0.211426     0.409549           0.365148   
min              0.000000            0.000000     0.000000           0.000000   
25%              0.000000            0.000000     0.000000           0.000000   
50%              0.000000            0.000000     0.000000           0.000000   
75%              0.000000            0.000000     0.000000           0.000000   
max              1.000000            1.000000     1.000000           1.000000   

             OTHER  ALBUMIN_MEDIAN  ALBUMIN_MEAN  ALBUMIN_MIN  ALBUMIN_MAX  \
count  1920.000000      821.000000    821.000000   821.000000   821.000000   
mean      0.809896        0.528527      0.528527     0.528527     0.528527   
std       0.392485        0.224100      0.224100     0.224100     0.224100   
min       0.000000       -1.000000     -1.000000    -1.000000    -1.000000   
25%       1.000000        0.605263      0.605263     0.605263     0.605263   
50%       1.000000        0.605263      0.605263     0.605263     0.605263   
75%       1.000000        0.605263      0.605263     0.605263     0.605263   
max       1.000000        1.000000      1.000000     1.000000     1.000000   

       ALBUMIN_DIFF  BE_ARTERIAL_MEDIAN  BE_ARTERIAL_MEAN  BE_ARTERIAL_MIN  \
count         821.0          821.000000        821.000000       821.000000   
mean           -1.0           -0.963433         -0.963433        -0.963433   
std             0.0            0.160870          0.160870         0.160870   
min            -1.0           -1.000000         -1.000000        -1.000000   
25%            -1.0           -1.000000         -1.000000        -1.000000   
50%            -1.0           -1.000000         -1.000000        -1.000000   
75%            -1.0           -1.000000         -1.000000        -1.000000   
max            -1.0            1.000000          1.000000         1.000000   

       BE_ARTERIAL_MAX  BE_ARTERIAL_DIFF  BE_VENOUS_MEDIAN  BE_VENOUS_MEAN  \
count       821.000000             821.0        821.000000      821.000000   
mean         -0.963433              -1.0         -0.931121       -0.931121   
std           0.160870               0.0          0.169509        0.169509   
min          -1.000000              -1.0         -1.000000       -1.000000   
25%          -1.000000              -1.0         -1.000000       -1.000000   
50%          -1.000000              -1.0         -1

# Tratamento


## Nome das colunas

In [ ]:
#Substituindo espaço por underscore
df_inicial.columns = df_inicial.columns.str.replace(' ', '_')

In [ ]:
df_inicial.head()

PATIENT_VISIT_IDENTIFIER  AGE_ABOVE65 AGE_PERCENTIL  GENDER  \
0                         0            1          60th       0   
1                         0            1          60th       0   
2                         0            1          60th       0   
3                         0            1          60th       0   
4                         0            1          60th       0   

   DISEASE_GROUPING_1  DISEASE_GROUPING_2  DISEASE_GROUPING_3  \
0                 0.0                 0.0                 0.0   
1                 0.0                 0.0                 0.0   
2                 0.0                 0.0                 0.0   
3                 0.0                 0.0                 0.0   
4                 0.0                 0.0                 0.0   

   DISEASE_GROUPING_4  DISEASE_GROUPING_5  DISEASE_GROUPING_6  HTN  \
0                 0.0                 1.0                 1.0  0.0   
1                 0.0                 1.0                 1.0  0.0   
2                 0.0                 1.0                 1.0  0.0   
3                 0.0                 1.0                 1.0  0.0   
4                 0.0                 1.0                 1.0  0.0   

   IMMUNOCOMPROMISED  OTHER  ALBUMIN_MEDIAN  ALBUMIN_MEAN  ALBUMIN_MIN  \
0                0.0    1.0             NaN           NaN          NaN   
1                0.0    1.0             NaN           NaN          NaN   
2                0.0    1.0        0.605263      0.605263     0.605263   
3                0.0    1.0             NaN           NaN          NaN   
4                0.0    1.0        0.000000      0.000000     0.000000   

   ALBUMIN_MAX  ALBUMIN_DIFF  BE_ARTERIAL_MEDIAN  BE_ARTERIAL_MEAN  \
0          NaN           NaN                 NaN               NaN   
1          NaN           NaN                 NaN               NaN   
2     0.605263          -1.0           -1.000000         -1.000000   
3          NaN           NaN                 NaN               NaN   
4     0.000000          -1.0           -0.871658         -0.871658   

   BE_ARTERIAL_MIN  BE_ARTERIAL_MAX  BE_ARTERIAL_DIFF  BE_VENOUS_MEDIAN  \
0              NaN              NaN               NaN               NaN   
1              NaN              NaN               NaN               NaN   
2        -1.000000        -1.000000              -1.0         -1.000000   
3              NaN              NaN               NaN               NaN   
4        -0.871658        -0.871658              -1.0         -0.863874   

   BE_VENOUS_MEAN  BE_VENOUS_MIN  BE_VENOUS_MAX  BE_VENOUS_DIFF  \
0             NaN            NaN            NaN             NaN   
1             NaN            NaN            NaN             NaN   
2       -1.000000      -1.000000      -1.000000            -1.0   
3             NaN            NaN            NaN             NaN   
4       -0.863874      -0.863874      -0.863874            -1.0   

   BIC_ARTERIAL_MEDIAN  BIC_ARTERIAL_MEAN  BIC_ARTERIAL_MIN  BIC_ARTERIAL_MAX  \
0                  NaN                NaN               NaN               NaN   
1                  NaN                NaN               NaN               NaN   
2            -0.317073          -0.317073         -0.317073         -0.317073   
3                  NaN                NaN               NaN               NaN   
4            -0.317073          -0.317073         -0.317073         -0.317073   

   BIC_ARTERIAL_DIFF  BIC_VENOUS_MEDIAN  BIC_VENOUS_MEAN  BIC_VENOUS_MIN  \
0                NaN                NaN              NaN             NaN   
1                NaN                NaN              NaN             NaN   
2               -1.0          -0.317073        -0.317073       -0.317073   
3                NaN                NaN              NaN             NaN   
4               -1.0          -0.414634        -0.414634       -0.414634   

   BIC_VENOUS_MAX  BIC_VENOUS_DIFF  BILLIRUBIN_MEDIAN  BILLIRUBIN_MEAN  \
0             NaN              NaN                NaN   

## Tratamento de colunas vazias (ou NaN)

Como será feito o preenchimento/tratamento dos dados vazios/nulos.

1. Tendo em vista a necessidade de prever os dados o mais rápido possível, a janela alvo para a previsão e do tratamento será a primeira (window: "0-2").

2. Inicialmente será feito o preparo da primeira janela com o marcador da UTI verdadeiro, para todos os pacientes que vieram a precisar dela depois da primeira janela. Para que seja possível identificar que precisaram da UTI.

3. Agora para o preenchimento dos dados vazios, levando em consideração a descrição do proprio desafio no Kaggle, será usado o dado histórico do paciente, sendo utilizado o próximo valor válido, desconsiderando as janelas que já deram entrada na UTI (coluna ICU).

O que será feito na prática:

1. Através de uma função, será passado o valor `UCI = 1` para a primeira janela de todos os pacientes que tiveram que ir para a UTI em algum momento.

1. Será feito uma seleção dos indices (linhas) que tiverem as condições: janela diferente da primeira `WINDOW != '0-2'` & marcador de UTI verdadeiro `ICU == 1`.

2. Serão removidas as linhas que tenham as condições citadas anteriormente, e será usado o metodo backfill `pandas.DataFrame.fillna(method=bfill)`

3. Por ultimo serão removidos os pacientes que ficarem ficaram com dados nulos (vazios) depois de todo o tratamento,

In [ ]:
df = df_inicial

In [ ]:
def preparando_janela(grupo):
    if any(grupo['ICU']):
        grupo.loc[grupo['WINDOW'] == '0-2', 'ICU'] = 1
    return grupo

In [ ]:
df = df.groupby('PATIENT_VISIT_IDENTIFIER').apply(preparando_janela)

In [ ]:
remover_linhas = df.query('WINDOW != "0-2" & ICU == 1').index.tolist()

In [ ]:
df = df.drop(index=remover_linhas)

In [ ]:
df.update(df.groupby('PATIENT_VISIT_IDENTIFIER', as_index=False).fillna(method='bfill', axis=0))

In [ ]:
df = df.query('WINDOW == "0-2"')

In [ ]:
df = df.dropna()

In [ ]:
df.describe()

PATIENT_VISIT_IDENTIFIER  AGE_ABOVE65      GENDER  DISEASE_GROUPING_1  \
count                301.000000   301.000000  301.000000          301.000000   
mean                 196.152824     0.415282    0.385382            0.093023   
std                  110.992026     0.493591    0.487496            0.290949   
min                    0.000000     0.000000    0.000000            0.000000   
25%                  106.000000     0.000000    0.000000            0.000000   
50%                  197.000000     0.000000    0.000000            0.000000   
75%                  294.000000     1.000000    1.000000            0.000000   
max                  384.000000     1.000000    1.000000            1.000000   

       DISEASE_GROUPING_2  DISEASE_GROUPING_3  DISEASE_GROUPING_4  \
count          301.000000          301.000000          301.000000   
mean             0.023256            0.083056            0.019934   
std              0.150966            0.276427            0.140005   
min              0.000000            0.000000            0.000000   
25%              0.000000            0.000000            0.000000   
50%              0.000000            0.000000            0.000000   
75%              0.000000            0.000000            0.000000   
max              1.000000            1.000000            1.000000   

       DISEASE_GROUPING_5  DISEASE_GROUPING_6         HTN  IMMUNOCOMPROMISED  \
count          301.000000          301.000000  301.000000         301.000000   
mean             0.112957            0.046512    0.166113           0.159468   
std              0.317067            0.210941    0.372802           0.366722   
min              0.000000            0.000000    0.000000           0.000000   
25%              0.000000            0.000000    0.000000           0.000000   
50%              0.000000            0.000000    0.000000           0.000000   
75%              0.000000            0.000000    0.000000           0.000000   
max              1.000000            1.000000    1.000000           1.000000   

            OTHER  ALBUMIN_MEDIAN  ALBUMIN_MEAN  ALBUMIN_MIN  ALBUMIN_MAX  \
count  301.000000      301.000000    301.000000   301.000000   301.000000   
mean     0.810631        0.583843      0.583843     0.583843     0.583843   
std      0.392454        0.120264      0.120264     0.120264     0.120264   
min      0.000000       -0.578947     -0.578947    -0.578947    -0.578947   
25%      1.000000        0.605263      0.605263     0.605263     0.605263   
50%      1.000000        0.605263      0.605263     0.605263     0.605263   
75%      1.000000        0.605263      0.605263     0.605263     0.605263   
max      1.000000        1.000000      1.000000     1.000000     1.000000   

       ALBUMIN_DIFF  BE_ARTERIAL_MEDIAN  BE_ARTERIAL_MEAN  BE_ARTERIAL_MIN  \
count         301.0          301.000000        301.000000       301.000000   
mean           -1.0           -0.999147         -0.999147        -0.999147   
std             0.0            0.012565          0.012565         0.012565   
min            -1.0           -1.000000         -1.000000        -1.000000   
25%            -1.0           -1.000000         -1.000000        -1.000000   
50%            -1.0           -1.000000         -1.000000        -1.000000   
75%            -1.0           -1.000000         -1.000000        -1.000000   
max            -1.0           -0.786096         -0.786096        -0.786096   

       BE_ARTERIAL_MAX  BE_ARTERIAL_DIFF  BE_VENOUS_MEDIAN  BE_VENOUS_MEAN  \
count       301.000000             301.0        301.000000      301.000000   
mean         -0.999147              -1.0         -0.967125       -0.967125   
std           0.012565               0.0          0.097800        0.097800   
min          -1.000000              -1.0         -1.000000       -1.000000   
25%          -1.000000              -1.0         -1.000000       -1.000000   
50%          -1.000000              -1.0         -1.000000       -1.000000   


Depois de todo o preenchimento e tratamento, restaram 301 pacientes

## Segregação dos tipos de dados




## Trabalhar correlação entre colunas para remover colunas com alta correlação

## Transformar em dados categoricos o "AGE_PERCENTIL"

# Salvar o dataframe em CSV no Repositório do Github